### PR TITLE
Adding ListPlugins wrapper

### PR DESCRIPTION
**Pull request by yunhee-l**
### Summary
Adding wrapper method for go-dockerclient's ListPlugins method, and a convenience method (ListPluginsWithFilters) to apply search filters (see https://docs.docker.com/engine/reference/commandline/plugin_ls/#filtering). go-dockerclient's ListPlugins method currently does not handle filters, but when we or someone else submits PR for the fix we will refactor these methods.

### Implementation details
ListPlugins wrapper method and ListPluginsWithFilters on top of it.

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [ ] Integration tests on Linux (`make run-integ-tests`) pass
- [ ] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [ ] Functional tests on Linux (`make run-functional-tests`) pass
- [ ] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/engine/docker_client_test.go [hunk 1]
```diff
@@ -1435,3 +1435,104 @@ func TestRemoveVolume(t *testing.T) {
...
+	wait.Add(1)
+	mockDocker.EXPECT().ListPlugins(gomock.Any()).Do(func(x interface{}) {
+		warp <- time.Now()
+		wait.Wait()
+		// Don't return, verify timeout happens
+		// TODO remove the MaxTimes by cancel the context passed to CreateVolume
+		// when issue #1212 is resolved
+	}).MaxTimes(1)
+	response := client.ListPlugins(timeout)
+	assert.Error(t, response.Error, "expected error for timeout")
```
Comment (petderek) [inline agent/engine/docker_client_test.go]: I think you could replace these assertions with [EqualError](https://godoc.org/github.com/stretchr/testify/assert#EqualError).

File: agent/engine/docker_client.go [hunk 2]
```diff
@@ -1134,6 +1145,85 @@ func (dg *dockerGoClient) removeVolume(ctx context.Context, name string) error {
...
+					filteredPluginNames = append(filteredPluginNames, pluginDetail.Name)
+					break
+				}
+			}
+		}
+	}
+	return filteredPluginNames, nil
+}
+
+func (dg *dockerGoClient) ListPlugins(timeout time.Duration) listPluginsResponse {
```
Comment (petderek) [inline agent/engine/docker_client.go]: What do you think about making timeout a property of the docker client struct? That should still make it reasonably simple from a testing perspective.

File: agent/engine/docker_client.go [hunk 3]
```diff
@@ -1134,6 +1145,85 @@ func (dg *dockerGoClient) removeVolume(ctx context.Context, name string) error {
...
+	// Injecting the 'timeout' makes it easier to write tests.
+	// Eventually, the context should be initialized from a parent root context
+	// instead of TODO.
+	ctx, cancel := context.WithTimeout(context.TODO(), timeout)
+	defer cancel()
+
+	// Buffered channel so in the case of timeout it takes one write, never gets
+	// read, and can still be GC'd
+	response := make(chan listPluginsResponse, 1)
+	go func() { response <- dg.listPlugins(ctx) }()
```
Comment (petderek) [inline agent/engine/docker_client.go]: Why does this need to happen asynchronously? As far as I can tell, we immediately wait on the goroutine we start.

File: agent/engine/docker_client_test.go [hunk 4]
```diff
@@ -1435,3 +1435,104 @@ func TestRemoveVolume(t *testing.T) {
...
+		Name: pluginName,
+		Tag: pluginTag,
+		Active: true,
+	}
+
+	mockDocker.EXPECT().ListPlugins(gomock.Any()).Return([]docker.PluginDetail{pluginDetail}, nil)
+
+	response := client.ListPlugins(inspectContainerTimeout)
+	assert.NoError(t, response.Error)
+	assert.Equal(t, pluginDetail.ID, response.Plugins[0].ID)
```
Comment (petderek) [inline agent/engine/docker_client_test.go]: assertEqual _should_ do an equality check on the whole struct... so one line should be fine?

File: agent/engine/docker_client_test.go [hunk 5]
```diff
@@ -1435,3 +1435,104 @@ func TestRemoveVolume(t *testing.T) {
...
+	assert.Equal(t, pluginDetail.Name, response.Plugins[0].Name)
+	assert.Equal(t, pluginDetail.Tag, response.Plugins[0].Tag)
+	assert.True(t, response.Plugins[0].Active)
+}
+
+func TestListPluginsWithFilter(t *testing.T) {
```
Comment (petderek) [inline agent/engine/docker_client_test.go]: Nice!

File: agent/engine/docker_client.go [hunk 6]
```diff
@@ -73,6 +75,7 @@ const (
 	removeContainerTimeout  = 5 * time.Minute
 	inspectContainerTimeout = 30 * time.Second
 	removeImageTimeout      = 3 * time.Minute
+	ListPluginsTimeout      = 1 * time.Minute
```
Comment (aaithal) [inline agent/engine/docker_client.go]: lint: needs a comment. You can move this to be with rest of the global const's as well

File: agent/engine/docker_client.go [hunk 7]
```diff
@@ -158,6 +161,14 @@ type DockerClient interface {
 	// RemoveVolume removes a volume by its name. A timeout value should be provided for the request
 	RemoveVolume(string, time.Duration) error
 
+	// ListPluginsWithFilters returns the set of docker plugins installed on the host, filtered by options provided.
+	// A timeout value should be provided for the request.
+	ListPluginsWithFilters(bool, []string, time.Duration) ([]string, error)
```
Comment (aaithal) [inline agent/engine/docker_client.go]: Just for my own understanding, what are the use-cases for invoking `ListPluginsWithFilters(false, ...)`?

Comment (yunhee-l) [reply agent/engine/docker_client.go]: this is where capability=volumedriver (and enabled=true) filter is supposed to be applied, but go-dockerclient's ListPlugins currently doesn't handle filters. I am going to submit PR for this fix but didn't want to block our work.

Comment (aaithal) [reply agent/engine/docker_client.go]: ah, makes sense. so, more for the sake of completion than anything else.

File: agent/engine/docker_client.go [hunk 8]
```diff
@@ -1134,6 +1145,85 @@ func (dg *dockerGoClient) removeVolume(ctx context.Context, name string) error {
...
 }
 
+type listPluginsResponse struct {
+	Plugins []docker.PluginDetail
+	Error   error
+}
+
+// ListPluginsWithFilters currently is a convenience method as go-dockerclient doesn't implement fitered list. When we or someone else submits
+// PR for the fix we will refactor this to pass in the fiters.
+func (dg *dockerGoClient) ListPluginsWithFilters(enabledFilter bool, capabilityFilter []string, timeout time.Duration) ([]string, error) {
```
Comment (aaithal) [inline agent/engine/docker_client.go]: nit: `enabled` and `capabilities` might be enough here for param names.

File: agent/engine/docker_client.go [hunk 9]
```diff
@@ -1134,6 +1145,85 @@ func (dg *dockerGoClient) removeVolume(ctx context.Context, name string) error {
 	return nil
 }
 
+type listPluginsResponse struct {
+	Plugins []docker.PluginDetail
+	Error   error
+}
+
+// ListPluginsWithFilters currently is a convenience method as go-dockerclient doesn't implement fitered list. When we or someone else submits
```
Comment (sharanyad) [inline agent/engine/docker_client.go]: Comments explaining what the arguments do here might be helpful for readability.

File: agent/engine/docker_client.go [hunk 10]
```diff
@@ -1134,6 +1145,85 @@ func (dg *dockerGoClient) removeVolume(ctx context.Context, name string) error {
...
+
+	var filteredPluginNames []string
+	response := dg.ListPlugins(timeout)
+
+	if response.Error != nil {
+		return nil, response.Error
+	}
+
+	for _, pluginDetail := range response.Plugins {
+		if pluginDetail.Active != enabledFilter {
```
Comment (sharanyad) [inline agent/engine/docker_client.go]: What does `pluginDetail.Active` denote here? Don't we always expect the plugin to be active?

File: agent/engine/docker_client_test.go (hunk 1 above)
Comment (yunhee-l) [reply agent/engine/docker_client_test.go]: this checks for the error message, and for timeout errors it's "Could not transition to ${err.transition}; timed out after waiting ${err.duration}". see https://github.com/aws/amazon-ecs-agent/blob/master/agent/engine/errors.go#L59
I'll keep this as is, it's confusing to mention transition here (which we should fix, not all timeouts relate to transitioning container).

File: agent/engine/docker_client.go (hunk 2 above)
Comment (yunhee-l) [reply agent/engine/docker_client.go]: I initially did, but looking at our DockerClient interface, all timeouts for APIs are passed in as a param. I don't know the design decision behind this, so I just went with being consistent. Given that we are using constants for these timeouts, maybe we could refactor it for all APIs - but let's do that as a separate PR.

File: agent/engine/docker_client.go (hunk 3 above)
Comment (yunhee-l) [reply agent/engine/docker_client.go]: it was the pattern for all docker API calls (or so I was told). I couldn't find out whether plugin ls was a synchronous or asynchronous call though.

File: agent/engine/docker_client_test.go (hunk 4 above)
Comment (yunhee-l) [reply agent/engine/docker_client_test.go]: Fixed.

File: agent/engine/docker_client.go (hunk 6 above)
Comment (yunhee-l) [reply agent/engine/docker_client.go]: Where do we keep global constants?

File: agent/engine/docker_client.go (hunk 8 above)
Comment (yunhee-l) [reply agent/engine/docker_client.go]: Fixed

File: agent/engine/docker_client.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (yunhee-l) [reply agent/engine/docker_client.go]: this is arbitrary, but given the fact that listPlugins basically returns the contents of in-memory map (see https://tinyurl.com/yaep53uw) it shouldn't need to be too long.

File: agent/engine/docker_client.go (hunk 10 above)
Comment (yunhee-l) [reply agent/engine/docker_client.go]: it is the same as "Enabled". You can install a plugin and then disable it.

File: agent/engine/docker_client.go (hunk 9 above)
Comment (yunhee-l) [reply agent/engine/docker_client.go]: Fixed

**Pull request merged**